### PR TITLE
Update validation logic for kubenet network plugin.

### PR DESCRIPTION
**Pull request by metacpp**
1. After sync-up with AKS team, when network plugin is `kubenet`, `dockerBridgeCidr`, `dnsServiceIP `, and `serviceCidr`.
2. Update the test case for `kubenet` scenario.
3. Update the documentation for `azure` and `kubenet`.

The result of related acceptance tests (8/8/2018):
```
=== RUN   TestAccAzureRMKubernetesCluster_importBasic
--- PASS: TestAccAzureRMKubernetesCluster_importBasic (1347.68s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_basic
--- PASS: TestAccAzureRMKubernetesCluster_basic (1466.90s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_internalNetwork
--- PASS: TestAccAzureRMKubernetesCluster_internalNetwork (1505.84s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_advancedNetworkingAzureComplete
--- PASS: TestAccAzureRMKubernetesCluster_advancedNetworkingAzureComplete (1509.45s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_advancedNetworkingAzure
--- PASS: TestAccAzureRMKubernetesCluster_advancedNetworkingAzure (1509.38s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_advancedNetworkingKubenet
--- PASS: TestAccAzureRMKubernetesCluster_advancedNetworkingKubenet (1520.81s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_advancedNetworkingKubenetComplete
--- PASS: TestAccAzureRMKubernetesCluster_advancedNetworkingKubenetComplete (1566.04s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_addAgent
--- PASS: TestAccAzureRMKubernetesCluster_addAgent (1966.13s)
PASS
=== RUN   TestAccAzureRMKubernetesCluster_upgradeConfig
--- PASS: TestAccAzureRMKubernetesCluster_upgradeConfig (2089.39s)
PASS
ok    ../azurerm/test-azurerm    2089.524s
```

### CODE REVIEW

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 1]
```diff
@@ -23,30 +23,6 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
-				// then ensure the conditionally-required fields are set
-				profile := rawProfiles[0].(map[string]interface{})
-				networkPlugin := profile["network_plugin"].(string)
-
-				if networkPlugin == "kubenet" {
-					dockerBridgeCidr := profile["docker_bridge_cidr"].(string)
-					dnsServiceIP := profile["dns_service_ip"].(string)
-					serviceCidr := profile["service_cidr"].(string)
-
-					if dockerBridgeCidr == "" || dnsServiceIP == "" || serviceCidr == "" {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster.go]: we should add this validation back in but ensure that if one of the fields is set then the other fields are set too

File: azurerm/resource_arm_kubernetes_cluster_test.go [hunk 2]
```diff
@@ -526,9 +523,6 @@ resource "azurerm_kubernetes_cluster" "test" {
 	
   network_profile {
     network_plugin     = "kubenet"
-    dns_service_ip     = "10.10.0.10"
-    docker_bridge_cidr = "172.18.0.1/16"
-    service_cidr       = "10.10.0.0/16"
   }
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster_test.go]: we should update to make two tests here, one for when this is not specified and one for when this is

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 3]
```diff
@@ -217,19 +217,19 @@ The following arguments are supported:
...
 
-* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
+* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. Changing this forces a new resource to be created.
 
 ~> **NOTE:** This range should not be used by any network element on or connected to this VNet. Service address CIDR must be smaller than /12.
 
-* `dns_service_ip` - (Optional) IP address within the Kubernetes service address range that will be used by cluster service discovery (kube-dns). This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
+* `dns_service_ip` - (Optional) IP address within the Kubernetes service address range that will be used by cluster service discovery (kube-dns). Changing this forces a new resource to be created.
 
-* `docker_bridge_cidr` - (Optional) IP address (in CIDR notation) used as the Docker bridge IP address on nodes. This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/kubernetes_cluster.html.markdown]: we should add in "This field can only be set when `network_plugin` is set to `kubenet`"

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 4]
```diff
@@ -217,19 +217,19 @@ The following arguments are supported:
...
+* `network_plugin` - (Required) Network plugin to use for networking. Currently supported values are `azure` and `kubenet`. Changing this forces a new resource to be created.
 
 -> **NOTE:** When `network_plugin` is set to `azure` - the `vnet_subnet_id` field in the `agent_pool_profile` block must be set.
 
-* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
+* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. Changing this forces a new resource to be created.
 
 ~> **NOTE:** This range should not be used by any network element on or connected to this VNet. Service address CIDR must be smaller than /12.
 
-* `dns_service_ip` - (Optional) IP address within the Kubernetes service address range that will be used by cluster service discovery (kube-dns). This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/kubernetes_cluster.html.markdown]: we should add in "This field can only be set when `network_plugin` is set to `kubenet`"

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 5]
```diff
@@ -217,19 +217,19 @@ The following arguments are supported:
 
 `network_profile` supports the following:
 
-* `network_plugin` - (Required) Network plugin to use for networking. Currently supported values are 'azure' and 'kubenet'. Changing this forces a new resource to be created.
+* `network_plugin` - (Required) Network plugin to use for networking. Currently supported values are `azure` and `kubenet`. Changing this forces a new resource to be created.
 
 -> **NOTE:** When `network_plugin` is set to `azure` - the `vnet_subnet_id` field in the `agent_pool_profile` block must be set.
 
-* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. This is required when `network_plugin` is set to `kubenet`. Changing this forces a new resource to be created.
+* `service_cidr` - (Optional) The Network Range used by the Kubernetes service. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/kubernetes_cluster.html.markdown]: we should add in "This field can only be set when `network_plugin` is set to `kubenet`"

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 6]
```diff
@@ -23,30 +23,6 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
-				rawProfiles := v.([]interface{})
-				if len(rawProfiles) == 0 {
-					return nil
-				}
-
-				// then ensure the conditionally-required fields are set
-				profile := rawProfiles[0].(map[string]interface{})
-				networkPlugin := profile["network_plugin"].(string)
-
-				if networkPlugin == "kubenet" {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster.go]: we can also invert this logic to ensure they're not set when the networkPlugin is "azure"

Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster.go]: Both `azure` and `kubenet` need to be taken care.

---
In reply to: [207462580](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207462580) [](ancestors = 207462580)

File: azurerm/resource_arm_kubernetes_cluster.go (hunk 1 above)
Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster.go]: They either need to be all empty or all non-empty, this is server side logic. Although we can add  it here to improve the experience, but we also introduced the risk to guess the server logic. Anyway, I made the change as you like.

---
In reply to: [207461949](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207461949) [](ancestors = 207461949)

File: azurerm/resource_arm_kubernetes_cluster_test.go (hunk 2 above)
Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster_test.go]: Yeah, 2 legacy tests will use different settings: one is all empty, the other is all non-empty.

---
In reply to: [207462036](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207462036) [](ancestors = 207462036)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 5 above)
Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: This field can only bet set with other 2 fields.

---
In reply to: [207462162](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207462162) [](ancestors = 207462162)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 4 above)
Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: Updated.

---
In reply to: [207462154](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207462154) [](ancestors = 207462154)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 3 above)
Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: Updated.

---
In reply to: [207462137](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207462137) [](ancestors = 207462137)

**Comment by metacpp** [comment]
@jeffreyCline is added to the review. #Closed

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 7]
```diff
@@ -34,13 +35,14 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
+				if networkPlugin == "kubenet" || networkPlugin == "azure" {
 					dockerBridgeCidr := profile["docker_bridge_cidr"].(string)
 					dnsServiceIP := profile["dns_service_ip"].(string)
 					serviceCidr := profile["service_cidr"].(string)
 
-					if dockerBridgeCidr == "" || dnsServiceIP == "" || serviceCidr == "" {
-						return fmt.Errorf("If the `network_plugin` is set to `kubenet` then the fields `docker_bridge_cidr`, `dns_service_ip` and `service_cidr` must not be empty.")
+					if !((dockerBridgeCidr == "" && dnsServiceIP == "" && serviceCidr == "") ||
+						(dockerBridgeCidr != "" && dnsServiceIP != "" && serviceCidr != "")) {
+						return fmt.Errorf("`docker_bridge_cidr`, `dns_service_ip` and `service_cidr` should all be empty or neither should be empty.")
```
Comment (jeffreyCline) [inline azurerm/resource_arm_kubernetes_cluster.go]: This is a confusing error msg, could we change it to something like:
`docker_bridge_cidr`, `dns_service_ip` and `service_cidr` should all be empty or all should be set.

Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster.go]: Good suggestion, updated it.

---
In reply to: [207693716](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207693716) [](ancestors = 207693716)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 3 above)
Comment (evillgenius75) [reply website/docs/r/kubernetes_cluster.html.markdown]: It should also be noted that the docker_bridge_cidr must have the last octet noted or the RP will fail. e.g 172.26.0.1/16 not 172.26.0.0/16 The server side logic is testing for this last octet and will stop the deployment.

Also the logic above for dns_service_ip is incorrect. It is only required is serviceCidr is entered not docker_bride_cidr

File: azurerm/resource_arm_kubernetes_cluster_test.go [hunk 8]
```diff
@@ -410,7 +441,9 @@ resource "azurerm_kubernetes_cluster" "test" {
 `, rInt, location, rInt, rInt, rInt, clientId, clientSecret)
 }
 
-func testAccAzureRMKubernetesCluster_advancedNetworkingKubenet(rInt int, clientId string, clientSecret string, location string) string {
+func testAccAzureRMKubernetesCluster_advancedNetworking(rInt int, clientId string, clientSecret string, location string,
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster_test.go]: given the values are going to be the same for the `dnsServiceIp `, `dockerBridgeCidr ` and `serviceCidr` fields in both Kubenet and Azure in both cases - we can duplicate the test configuration and hard-code the values for both cases?

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 9]
```diff
@@ -34,13 +35,14 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
-				if networkPlugin == "kubenet" {
+				if networkPlugin == "kubenet" || networkPlugin == "azure" {
 					dockerBridgeCidr := profile["docker_bridge_cidr"].(string)
 					dnsServiceIP := profile["dns_service_ip"].(string)
 					serviceCidr := profile["service_cidr"].(string)
 
-					if dockerBridgeCidr == "" || dnsServiceIP == "" || serviceCidr == "" {
-						return fmt.Errorf("If the `network_plugin` is set to `kubenet` then the fields `docker_bridge_cidr`, `dns_service_ip` and `service_cidr` must not be empty.")
+					if !((dockerBridgeCidr == "" && dnsServiceIP == "" && serviceCidr == "") ||
+						(dockerBridgeCidr != "" && dnsServiceIP != "" && serviceCidr != "")) {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster.go]: _minor / not a blocker_ we could probably make this more readable by doing:
```
hasDockerCidr := dockerBridgeCidr != ""
hasServiceIP := dnsServiceIP != ""
hasServiceCidr := serviceCidr != ""
allEmpty := !hasDockerCidr && !hasServiceIP && !hasServiceCidr
allSet := hasDockerCidr && hasServiceIP && hasServiceCidr
if !allEmpty && !allSet {
  return fmt.Errorf("...")
}
```

File: azurerm/resource_arm_kubernetes_cluster_test.go (hunk 8 above)
Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster_test.go]: Thanks for the suggestion.The extra parameters provide functionalities to add more testing combinations, besides all empty or all set. I would like to leave it here.

---
In reply to: [207946292](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207946292) [](ancestors = 207946292)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 3 above)
Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: We tried this combination:
```hcl
network_profile {
    network_plugin     = "kubenet"
    dns_service_ip     = "10.10.0.10"
    docker_bridge_cidr = ""
    service_cidr       = "10.10.0.0/16"
}
```
It still fails due to not all of 3 are set.

---
In reply to: [207924196](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207924196) [](ancestors = 207924196)

Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: For the octet, I don't quite understand, can you be more specific on the documentation change?

---
In reply to: [208059432](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r208059432) [](ancestors = 208059432,207924196)

File: azurerm/resource_arm_kubernetes_cluster.go (hunk 9 above)
Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster.go]: Good suggestion, let me figure out a better way.

---
In reply to: [207947438](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r207947438) [](ancestors = 207947438)

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 3 above)
Comment (evillgenius75) [reply website/docs/r/kubernetes_cluster.html.markdown]: If someone enters what they normally think of a CIDR address is such as 172.16.0.0/16 the RP will kick back an error saying that docker_bridge_Cidr is not a valid cidr address. It must be the first available address in the space and the prefix number, not the network name and prefix.

Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: Where can we find the these rules? I feel we probably can provide a link to how to define the addresses here instead. Anyway, feel free to send another PR to update this document, that will be more efficient.

---
In reply to: [208078669](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r208078669) [](ancestors = 208078669)

Comment (metacpp) [reply website/docs/r/kubernetes_cluster.html.markdown]: Let's take it offline.

---
In reply to: [208059738](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r208059738) [](ancestors = 208059738,208059432,207924196)

File: azurerm/resource_arm_kubernetes_cluster_test.go (hunk 8 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_kubernetes_cluster_test.go]: @metacpp this should be refactored to match the other resources, since the values intentionally shouldn't be set to empty strings if they're unset

Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster_test.go]: I did the refactoring to align with your preference here.

---
In reply to: [208550336](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1715#discussion_r208550336) [](ancestors = 208550336)

**Comment by tombuildsstuff** [comment]
@metacpp heads up that I've pushed a commit to refactor the test case to match the other test cases

**Comment by tombuildsstuff** [comment]
Tests pass:

screenshot 2018-08-09 at 16 23 44
